### PR TITLE
Implement SPICE-0009 External Readers

### PR DESCRIPTION
**Pull request by HT154**
[SPICE-0009](https://github.com/apple/pkl-evolution/pull/10)

Includes the changes from https://github.com/apple/pkl/pull/640

Adoption of this feature in language binding libraries:
* https://github.com/apple/pkl-go/pull/101
* https://github.com/apple/pkl-swift/pull/26

### CODE REVIEW

File: pkl-server/src/test/kotlin/org/pkl/server/MessagePackCodecTest.kt
Comment (HT154) [inline pkl-server/src/test/kotlin/org/pkl/server/MessagePackCodecTest.kt:41]: Lost ~4 hours of debugging this suite hanging to this diff 😅

File: pkl-core/src/main/java/org/pkl/core/externalProcess/ExternalReaderRuntime.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/externalProcess/ExternalReaderRuntime.java:33]: Let's move this and all related "runtime" classes to pkl-config-java. These are not core concerns of Pkl, and we don't need to ship these classes with pkl-cli, for example.

File: pkl-core/src/main/java/org/pkl/core/messaging/MessageTransports.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/messaging/MessageTransports.java:58]:
```suggestion
        throw new IOException("external read failure: " + e.getMessage(), e.getCause());
```

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java:789]: Let's move class `Resolver` to its own class, and use it both here and for the resource reader (see `FileResolver`, `PackageResolver`, etc).

File: pkl-server/src/main/kotlin/org/pkl/server/Server.kt
Comment (bioball) [inline pkl-server/src/main/kotlin/org/pkl/server/Server.kt:260]: Do these external readers ever need to care about what the evaluator ID is?

File: stdlib/EvaluatorSettings.pkl
Comment (bioball) [inline stdlib/EvaluatorSettings.pkl:164]:
```suggestion

```

File: stdlib/EvaluatorSettings.pkl
Comment (bioball) [inline stdlib/EvaluatorSettings.pkl:160]:
```suggestion
  ///
  ///   * An absolute path
```

File: stdlib/EvaluatorSettings.pkl
Comment (bioball) [inline stdlib/EvaluatorSettings.pkl:98]:
```suggestion
/// Configuration for external module reader processes.
@Since { version = "0.27.0" }
externalModuleReaders: Mapping<String, ExternalReader>?

/// Configuration for external resource reader processes.
@Since { version = "0.27.0" }
externalResourceReaders: Mapping<String, ExternalReader>?
```

File: stdlib/EvaluatorSettings.pkl
Comment (bioball) [inline stdlib/EvaluatorSettings.pkl:154]:
```suggestion
@Since { version = "0.27.0" }
class ExternalReader {
```

**Review by bioball — CHANGES_REQUESTED** [review]
Did a first pass.

This is looking pretty good! Looks pretty polished, with just a couple of issues.

Design question: what do we want to do about env vars or CWD of the process?

What should users expect if thay do:

```
pkl --env-var foo=bar --external-module-reader ldap=my-ldap-reader
```

Should `my-ldap-reader` receive `foo=bar` as an env var? Or should there be some way to configure the env vars that get passed to this reader?

Same question applies for `--working-dir`.

Thoughts: probably not appropriate to pass these values into the process, but it should be clearly documented. But, we should probably have some way to configure env var, as well as CWD for these processes.

File: pkl-server/src/main/kotlin/org/pkl/server/Server.kt
Comment (HT154) [reply pkl-server/src/main/kotlin/org/pkl/server/Server.kt:260]: Not really, since the `MessageTransport` is distinct from the server's. One reason to still populate this is `PKL_DEBUG` logging. Because the message contents are printed, this could help debug issues in multi-evaluator scenarios.

File: pkl-core/src/main/java/org/pkl/core/externalProcess/ExternalReaderRuntime.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/externalProcess/ExternalReaderRuntime.java:33]: This class is used in the pkl-core tests (pkl-core/src/test/kotlin/org/pkl/core/externalProcess/TestExternalProcess.kt). I'm fine moving it if the pkl-core tests can depend on pkl-config-java.

File: pkl-core/src/main/java/org/pkl/core/externalProcess/ExternalProcess.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/externalProcess/ExternalProcess.java:25]: Any object to `ExternalReaderProcess`? Maybe I'm overthinking, but this seems more like an abstraction for a process pertaining to external readers than a reader abstraction pertaining to external processes.

Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/externalProcess/ExternalProcess.java:25]: `ExternalReaderProcess` works with me too!

File: pkl-core/src/main/java/org/pkl/core/messaging/Messages.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:519]: I won't block the PR on this, because it's a lot of cruft, but: we should make all of these records instead.

And the message interface getters should have the `get` prefix removed; e.g. `long requestId()` instead of `long getRequestId()` so that the records can conform to the interface easily.

We can leave this for a later refactor, though.

File: pkl-core/src/main/java/org/pkl/core/externalReader/ExternalReaderProcess.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/externalReader/ExternalReaderProcess.java:46]: It's valid to want to keep the processes alive between different evaluators, isn't it? It seems fine to want to close the evaluator, but keep the reader alive and pass it to a new evaluator.

```suggestion
   * <p>The {@link
```

File: pkl-core/src/main/java/org/pkl/core/externalReader/ExternalReaderProcessImpl.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/externalReader/ExternalReaderProcessImpl.java:42]: Nit: use `java.time.Duration` instead; e.g. `Duration.ofSeconds(3)`.

File: pkl-core/src/main/java/org/pkl/core/externalReader/ExternalReaderProcessImpl.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/externalReader/ExternalReaderProcessImpl.java:140]: Should be fine to close the transport at this point too? This will remove any response handlers.

```suggestion
      getTransport().send(new CloseExternalProcess());
      getTransport().close();
```

File: pkl-core/src/main/java/org/pkl/core/messaging/MessageTransports.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/messaging/MessageTransports.java:122]: Don't think this comment is super relevant anymore.

```suggestion
```

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java:267]: These factories should be thread-safe.
```suggestion
    private synchronized ExternalModuleResolver getResolver() throws ExternalReaderProcessException {
```

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java:259]:
```suggestion
    @GuardedBy("this")
    private ExternalModuleResolver resolver;
```

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java:781]:
```suggestion
  public static class External implements ModuleKey {
```

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java:138]:
```suggestion
  public static ModuleKey external(
      URI uri, ModuleReaderSpec spec, ExternalModuleResolver resolver) {
    return new External(uri, spec, resolver);
  }
```

File: pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java:685]: For consistency; no other readers have `Reader` name postfixes.

```suggestion
  private static final class MessageTransport implements ResourceReader {
```

File: stdlib/EvaluatorSettings.pkl
Comment (bioball) [inline stdlib/EvaluatorSettings.pkl:168]:
```suggestion

  /// Additional command line arguments passed to the external reader process.
```

**Review by bioball — CHANGES_REQUESTED** [review]
One more pass. Getting close!

File: pkl-config-java/src/main/java/org/pkl/config/java/ExternalReaderRuntime.java
Comment (HT154) [reply pkl-config-java/src/main/java/org/pkl/config/java/ExternalReaderRuntime.java:30]: Haha this actually goes back to my original proposal of a new pkl-msgapi project! Agreed on, if we eventually decide to publish this, having these bits be standalone and on doing that as followup work.

File: pkl-core/src/main/java/org/pkl/core/externalReader/package-info.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/externalReader/package-info.java:2]: `externalreader` sounds good to me. I was following the lead of `evaluatorSettings` 😅

File: pkl-core/src/main/java/org/pkl/core/messaging/Message.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Message.java:37]: I think this was to hard-differentiate between server messages and external reader messages, based on one of our discussions quite a while back. Definitely open to revising this.

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java:781]: Hmm, I had renamed this to align with `ResourceReaders.MessageTransport` since, like that class, this one is used by both external readers and client readers. I think I'd prefer to keep this name and preserve the symmetry. Thoughts?

File: pkl-core/src/main/java/org/pkl/core/messaging/Messages.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:117]: I took a look at what this would take and figured it was a small enough effort and a large enough breaking change to be worth doing now!

File: pkl-core/src/main/java/org/pkl/core/messaging/Message.java
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/messaging/Message.java:37]: Yeah, since many of these messages are shared between the two different flows, I'm not sure if it matters that much.

Maybe better to just keep this incremental.

Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Message.java]: Updated!

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java:781]: Ooh I see.

Maybe `ResourceReaders.ExternalResolver`?

And then call the other one `ResourceReaders.ExternalProcess`?

And then apply the same names on the module side?

**Review by bioball — APPROVED** [review]
Overall LGTM! Great work!

This also needs eyes from @holzensp @stackoverflow because it touches the core language.

**Review by stackoverflow — APPROVED** [review]
Looks good. Some small things.

File: pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java:59]: I though about this, but figured this approach (which matches the original Kotlin implementation) might have been chosen for efficiency reasons. This code is in a pretty hot path, so if the goal is to avoid allocations or iteration that would make sense to me. Happy to change this if there's consensus!

File: pkl-core/src/main/java/org/pkl/core/messaging/Messages.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: Maybe this is a tooling issue instead, but I wasn't able to get things working correctly with marking `byte[]` usage `@Nullable`. Some of the message passing APIs don't work correctly at all if you try to swap in empty arrays where `null` `ByteArray?` values were used previously.

Comment (stackoverflow) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: Maybe it's a message pack bug then, because Java will gladly allow you to set a `byte[]` to `null` and mark it `@Nullable`.

File: pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java
Comment (stackoverflow) [reply pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java:59]: I don't think performance should be a concern here. The JVM can easily optimize varargs parameters, specially on a hot path.
On the other hand I don't have strong opinions here, it was more of a nit.

File: pkl-core/src/main/java/org/pkl/core/messaging/Messages.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: I don't think it's msgpack-specific. This is what I'm seeing:

Screenshot 2024-10-24 at 9 36 27 AM

Comment (stackoverflow) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: Have you tested that's actually the case? It may be IntelliJ getting the analysis wrong.

Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: Yup, these appear to be legit:
Screenshot 2024-10-24 at 11 20 45 AM

```
> Task :pkl-core:compileJava FAILED
/Users/jbasch/src/pkl/pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackDecoder.java:122: error: illegal start of type
  protected static @Nullable []byte unpackByteArray(Map<Value, Value> map, String key) {
                             ^
/Users/jbasch/src/pkl/pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackDecoder.java:122: error: ';' expected
  protected static @Nullable []byte unpackByteArray(Map<Value, Value> map, String key) {
```

Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: Yeah, this appears to not play nice when bridging into Kotlin:
Screenshot 2024-10-24 at 11 51 25 AM

**Review by holzensp — APPROVED** [review]
I had some minor nits remaining, but they all got addressed ;) LGTM.

File: pkl-core/src/main/java/org/pkl/core/externalReader/package-info.java
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/externalReader/package-info.java]: Ah, I guess we're not the best at following Java conventions either 🫣

File: pkl-core/src/main/java/org/pkl/core/messaging/Messages.java
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: When annotating array types, the notation looks a little different. For `byte[]`, you would do `byte @Nullable []`. Java is pretty weird sometimes.

@stackoverflow is right--this should just be an array; boxing doesn't do anything special, and just using a byte array is better.

File: pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java:59]: I don't remember why the Kotlin side was written that way.

I think varargs would be slower than this because it would require a for loop (unless the JVM's JIT does some unrolling magic), but I don't think we'd have any perf issues even with a for loop.

Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java:59]: Let's just keep this; no need to rewrite something that works and is likely more performant.

File: pkl-core/src/main/java/org/pkl/core/messaging/Messages.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java:27]: Ah thank you! I struggled to find any hints toward this when I searched. Will get this updated shortly!

Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java]: Pushed this change, but I'm still getting build errors locally from what seems like the same issue. Everything checks out according to IntelliJ but compilation fails with
```
> Task :pkl-server:compileTestKotlin FAILED
e: .../pkl-server/src/test/kotlin/org/pkl/server/AbstractServerTest.kt: (199, 78): Null can not be a value of a non-null type ByteArray
e: .../pkl-server/src/test/kotlin/org/pkl/server/AbstractServerTest.kt: (227, 68): Null can not be a value of a non-null type ByteArray

> Task :pkl-core:compileTestKotlin FAILED
e: .../pkl-core/src/test/kotlin/org/pkl/core/externalreader/ExternalReaderRuntime.kt: (178, 19): Null can not be a value of a non-null type ByteArray
e: .../pkl-core/src/test/kotlin/org/pkl/core/externalreader/ExternalReaderRuntime.kt: (195, 70): Null can not be a value of a non-null type ByteArray
```

Any ideas? I've tried `gw clean` and a fresh gradle daemon and those didn't help.

Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java]: Hmm, weird. I'll take a look.

Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java]: I pushed a workaround that adds a constructor for `ReadResourceResponse` without `content`, but I'm really not in love with it.

Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java]: I _think_ this is a bug in the Kotlin compiler. What's also strange is that the analysis within IntelliJ is correct.

As a workaround for now, you can replace `null` with `byteArrayOf()`. We're planning on switching our annotations to JSpecify (https://github.com/apple/pkl/issues/698), so this issue might be naturally fixed once that happens.

Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java]: Hmm but one of these tests that passes null is explicitly doing so to test that it's handled correctly. Passing an empty byte array instead sort of defeats that. The null contents+error is a valid condition and is a result of how msgpack go encodes these messages.

Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java]: Ended up using `byteArrayOf()` and disabling the one test that relies specifically on null `content` with explanatory comments.

Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/messaging/Messages.java]: That's reasonable!

**Pull request merged**
